### PR TITLE
Fixed mono_type_get_name implementation

### PR DESCRIPTION
**Pull request by alexey-zakharov**
The PR changes the implementation of `mono_type_get_name` to return  type name including namespace.
This makes it consistent with mono implementation and the return value expectations on the Unity Editor side.

- `mono_type_get_name` is equivalent to `mono_type_get_name_full(type, MonoTypeNameFormat::MONO_TYPE_NAME_FORMAT_IL);`
- MONO_TYPE_NAME_FORMAT_IL is close to TypeString::FormatNamespace format flag and `Type.ToString()` 

I've also explored possibility of rerouting calls to the `Type.ToString()` and `Type.AssemblyQualifiedName` - it requires addressing of the string buffer disposal issue (it doesn't look like we can use `Marshal.FreeHGlobal` and `free` interchangeably?) as well as reviewing `scripting_*` api usage to accommodate .NET Runtime name formats. That is a potential next step in this api implementation

Note: I've decided to not implement MONO_TYPE_NAME_FORMAT_REFLECTION_QUALIFIED case to ensure we try to move towards to the public .NET API when interpreting type names

### CODE REVIEW

File: unity/embed_api_tests/main.cpp [hunk 1]
```diff
@@ -763,11 +772,11 @@ TEST(can_get_type_of_generic_field)
     gpointer ptr = nullptr;
     GET_AND_CHECK(field, mono_class_get_fields(klass, &ptr));
     CHECK(strcmp("genericField", mono_field_get_name(field)) == 0);
-    CHECK(strcmp(g_Mode == CoreCLR ? "!0" : "T", mono_type_get_name(mono_field_get_type(field))) == 0);
+    CHECK(strcmp("T", mono_type_get_name(mono_field_get_type(field))) == 0);
```
Comment (alexey-zakharov) [inline unity/embed_api_tests/main.cpp:775]: I'm not aware if the previous behavior (!0) was intentional and as I didn't experience the differences yet I hope that it should not matter

**Comment by UnityAlex** [comment]
> Note: I've decided to not implement MONO_TYPE_NAME_FORMAT_REFLECTION_QUALIFIED case to ensure we try to move towards to the public .NET API when interpreting type names

Yes this was/is a limitation of our current serialization system and the folks on that team didn't want to use the fully qualified type name as the string they use.

**Review by UnityAlex — APPROVED** [review]

**Comment by alexey-zakharov** [comment]
> Yes this was/is a limitation of our current serialization system and the folks on that team didn't want to use the fully qualified type name as the string they use.

good to know, thanks Alex! I expect this may break some of deserialization code, but probably it is ok to see if we can have it working in conjunction with serialization.
It is possible to have very close behavior with native flags, but if we want to move to C#, then there is no equivalent API and we'll need to do a manual string composition which will be expensive IMHO

**Comment by joncham** [comment]
> (it doesn't look like we can use Marshal.FreeHGlobal and free interchangeably?

Correct, we should not assume this. I think @UnityAlex has PR touching on this as well. We need to ensure all APIs that return an allocated pointer are compatible with `mono_unity_g_free` implementation.

**Review by joncham — APPROVED** [review]

**Comment by alexey-zakharov** [comment]
> Correct, we should not assume this. I think @UnityAlex has PR touching on this as well. We need to ensure all APIs that return an allocated pointer are compatible with `mono_unity_g_free` implementation.

we can use a lowbit flag to dispatch to `free` or `FreeHGlobal` within `mono_unity_g_free`, but need to do this synchronously for converted to C# api calls

**Pull request merged**